### PR TITLE
Add voxin module

### PR DESCRIPTION
**Pull request by gcasse**
This module is compatible with Voxin >= 3.0.

### CODE REVIEW

**Comment by sthibaul** [comment]
Hello,

Errr, I don't think we want to replicate almost all of the ibmtts.c source code like this. We already have such a thing between espeak and espeak-ng, and that has already posed questions, that's not the way we want to see the spd code go. See for instance https://github.com/brailcom/speechd/pull/373 which you do not have in your version, while I guess you want it. Sharing the implementation really is what we want to make it maintainable without headaches.

I have now integrated various changes you have made, notably the more flexible language selection. While looking at the difference between your driver and the current ibmtts, I am still left with the question which is still unanswered:

“I don't understand why using voxGetVoices: AIUI voxin is just a forwarder for ibmtts, that is not actually necessary: the speech-dispatcher plugin can just directly link against ibmtts, and use ibmtts-provided functions. Here, won't the eciGetAvailableLanguages call simply just work already? If not yet, can't that be solved instead of introducing yet more complexity? AFAICT, voxGetVoices ends up just calling eciGetAvailableLanguages anyway, so I don't see the benefit of calling the former rather the latter, but see the drawback: it prevents from linking again ibmtts directly.”

Is the current ibmtts.c not already working with voxin 3.0? What poses problem?

File: src/modules/voxin.c [hunk 1]
```diff
@@ -0,0 +1,1955 @@
...
+	if (old_index == new_index) {
+		DBG(DBG_MODNAME "LEAVE %s, no change", __FUNCTION__);
+		return;
+	}
+
+	language = voices[new_index].lang;
+	region = voices[new_index].variant;
+
+	/* Fix locale name for French Canadian */
+	if (!strcmp(language, "ca") && !strcmp(region, "FR")) {
```
Comment (sthibaul) [inline src/modules/voxin.c:1887]: Also, ewwww...
AIUI it's just a bug in voxin that you fixed in 451ba54dff0c84f731aede9e045ed2a0b0831d9f

**Comment by sthibaul** [comment]
Just to be clear: we can build the same source file twice with different `-D` parameters so we can use ifdefs and keep just one source file to maintain, that's probably what we'll do with espeak vs espeak-ng.

**Comment by gcasse** [comment]
Hello,

Before 2016, sd_ibmtts directly called e.g. the
eciGetAvailableLanguages() function from the IBM TTS 32 bits library.

Since January 2016, libvoxin implements the ECI API to originally ease
the integration of the 32 bits IBM TTS binaries in a 64 bits environment
for various accessibility clients (sd_ibmtts, emacspeak outloud server,...).

Today, sd_voxin is built for x86_64, armh, aarch64 (ARM only with the
Vocalizer Embedded voices).

**Comment by gcasse** [comment]
The strcmp line has been removed from my speechd fork but unfortunately
remained in this pull request.
Good catch!

**Comment by sthibaul** [comment]
> Today, sd_voxin is built for x86_64, armh, aarch64 (ARM only with the Vocalizer Embedded voices).

Does this mean that there are arm versions of the ECI engine binary? Or is the original binary run through simulation?

**Comment by gcasse** [comment]
sd_voxin on ARM is used for the Vocalizer Embedded voices.

**Comment by sthibaul** [comment]
> sd_voxin on ARM is used for the Vocalizer Embedded voices.

I don't understand what this mean. Is "Vocalizer Embedded" another product that provides the ECI interface? Is it something completely different that voxin happens to know how to drive? Is it available on i386 / x86_64 as well?

**Comment by gcasse** [comment]
Yes, more info at voxin.oralux.net

**Comment by sthibaul** [comment]
> Yes

Yes to what? Both my questions? From what I could find out about Vocalizer, it is available on i386/x86_64, but does not expose the ECI interface.

Which leads me wondering: why not just have a Vocalizer spd module? What benefit do we get from going through a voxin daemon? We already have an `orca -> spd server -> spd module -> pulseaudio` pipeline, I don't see why adding yet another stage, with yet more API translation, between the oldie ECI interface that was not meant for modern use and imposes its own constraints, and the Vocalizer interface which has a proper modern API. Adding another stage also increases synthesis latency.

**Comment by sthibaul** [comment]
As an interim solution, I have merged the voxin module, but factorized with ibmtts to avoid the whole code duplication.

**Pull request closed without merging**